### PR TITLE
5.0 Faces Standard Converters with Generics

### PR DESCRIPTION
**Pull request by pizzi80**
https://github.com/jakartaee/faces/issues/1822

+ Basic Converters with Generics
+ Basic Converters optimized

### CODE REVIEW

**Comment by mnriem** [comment]
@pizzi80 What is the status of this PR?

**Comment by pizzi80** [comment]
The task was finalized on June 22 with the commit https://github.com/eclipse-ee4j/mojarra/pull/5306/commits/6590ff1cbd9c7da5101ce3a6868f1978f48e8a8e. 

However, due to numerous merged pull requests in the interim, 
I had to synchronize my fork, make adjustments to some imports, and alter the name. 
(last commit)

This was necessitated by the decision of the JSF API community to defer the usage of Generics to Faces 5+

**Comment by mnriem** [comment]
@pizzi80 So this PR can be closed out?

**Comment by pizzi80** [comment]
If the master is now the version 5.0,
I think that it could finally be merged

**Comment by mnriem** [comment]
@BalusC @arjantijms What do you think?

File: impl/src/main/java/com/sun/faces/util/Util.java [hunk 1]
```diff
@@ -118,6 +119,9 @@ public class Util {
     // Log instance for this class
     private static final Logger LOGGER = FacesLogger.APPLICATION.getLogger();
 
+    public static final String EMPTY_STRING = "";
+    public static final String SPACE_STRING = " ";
```
Comment (BalusC) [inline impl/src/main/java/com/sun/faces/util/Util.java:123]: Should go in `RIConstants`. The `EMPTY_STRING` is already represented by `NO_VALUE`.

File: impl/src/main/java/com/sun/faces/util/Util.java [hunk 2]
```diff
@@ -505,6 +509,32 @@ public static void notNullAttribute(String attributeName, Object attribute) {
...
+        notNullArgs(arg1,arg2,arg3);
+        Objects.requireNonNull(arg4);
+    }
+
+    public static void notNullArgs( Object arg1 , Object arg2 , Object arg3 , Object arg4 , Object arg5 ) {
+        notNullArgs(arg1,arg2,arg3,arg4);
+        Objects.requireNonNull(arg5);
+    }
+
+    public static void notNullArgs( Object... objets ) {
```
Comment (BalusC) [inline impl/src/main/java/com/sun/faces/util/Util.java:532]: objets -> objects

File: impl/src/main/java/com/sun/faces/util/Util.java [hunk 3]
```diff
@@ -505,6 +509,32 @@ public static void notNullAttribute(String attributeName, Object attribute) {
...
+        Objects.requireNonNull(objets);
```
Comment (BalusC) [inline impl/src/main/java/com/sun/faces/util/Util.java:533]: can be omitted, varargs are never null in first place

File: impl/src/main/java/com/sun/faces/util/Util.java [hunk 4]
```diff
@@ -505,6 +509,32 @@ public static void notNullAttribute(String attributeName, Object attribute) {
...
+        for ( Object obj : objets)
+            Objects.requireNonNull(obj);
```
Comment (BalusC) [inline impl/src/main/java/com/sun/faces/util/Util.java:535]: `Arrays.stream(objects).forEach(Objects::requireNonNull)` :)

File: impl/src/main/java/jakarta/faces/convert/BigDecimalConverter.java [hunk 5]
```diff
@@ -72,51 +75,34 @@ public class BigDecimalConverter implements Converter {
...
-        value = value.trim();
-        if (value.length() < 1) {
+        // If the specified value is null or blank, return null
+        if ( value == null || value.isBlank() ) {
             return null;
         }
 
         try {
-            return new BigDecimal(value);
+            return new BigDecimal(value.trim());
```
Comment (BalusC) [inline impl/src/main/java/jakarta/faces/convert/BigDecimalConverter.java:87]: Microoptimization comment: trim()+length() is more efficient than isBlank()+trim(). As performance is important this change should be reverted IMO. 

Try to make the PR as focused as possible as the ticket describes (introducing generics in faces converers).

File: impl/src/main/java/jakarta/faces/convert/EnumConverter.java [hunk 6]
```diff
@@ -66,27 +69,6 @@ public class EnumConverter implements Converter, PartialStateHolder {
...
-    }
-
-    /**
-     * Instantiates an enum converter with a class where enum constants are taken from.
-     *
-     * @param targetClass Class where the enum constants are taken from by the converter methods.
-     */
-    public EnumConverter(Class targetClass) {
-        this.targetClass = targetClass;
-    }
```
Comment (BalusC) [inline impl/src/main/java/jakarta/faces/convert/EnumConverter.java:87]: This is a direct API change. This is not allowed. You should create a ticket in faces project and discuss why this should be removed and then create PR to mark it deprecated and then only remove it in a next major version.

Please revert.

**Review by BalusC — CHANGES_REQUESTED** [review]

File: impl/src/main/java/com/sun/faces/util/Util.java (hunk 4 above)
Comment (pizzi80) [reply impl/src/main/java/com/sun/faces/util/Util.java:535]: I like a lot lambdas but I bet is cheaper to do a simple for ;)

File: impl/src/main/java/jakarta/faces/convert/BigDecimalConverter.java (hunk 5 above)
Comment (pizzi80) [reply impl/src/main/java/jakarta/faces/convert/BigDecimalConverter.java:87]: questionable... but ok ;)

probably it depends a lot on how many 
empty/blank values vs real values (with vs without empty chars...)
the webapp have to convert on the average

the `isBlank` algo is very fast and cheap 

the `trim` one involves 2 loops + clone operation

File: impl/src/main/java/jakarta/faces/convert/EnumConverter.java (hunk 6 above)
Comment (pizzi80) [reply impl/src/main/java/jakarta/faces/convert/EnumConverter.java:87]: ouch ... ok!
... I though it was an implementation choice due to old EL/Java version ...

**Comment by pizzi80** [comment]
@BalusC Thanks for taking the time to review my code .... 

I'll fix everything the next week

**Comment by pizzi80** [comment]
During this work I've found that inside the generic `NumberConverter`

there are some ugly hacks for Java 1.4 and Java 7 that use reflection
to invoke some methods that are obviously implemented in Java 8+ ... 

Jsf 2.3 is Java 8+, right?

Can I remove them on a separate PR?

**Comment by BalusC** [comment]
API has been moved into Faces project. Current PR is superseded by https://github.com/jakartaee/faces/pull/1942

**Pull request closed without merging**